### PR TITLE
zero lock validation

### LINKED ISSUE #999: add validation for size and lock on newallocation and updateallocation
- Add validation that allocation size cannot be updated to less than amount stored in allocation
- Add --lock 0 validation

related links
- zbox: https://github.com/0chain/zboxcli/pull/172
- gosdk: https://github.com/0chain/gosdk/pull/375
- 0chain: https://github.com/0chain/0chain/issues/999

### PR DESCRIPTION
**Pull request by Kishan-Dhakan**
## Fixes
- https://github.com/0chain/0chain/issues/999
- https://github.com/0chain/0chain/issues/997

## Tests 
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | fix/alloc-update  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | zero-lock-validation  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | fix/alloc-update  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | zero-lock-validation  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

File: code/go/0chain.net/smartcontract/storagesc/allocation.go [hunk 1]
```diff
@@ -1080,14 +1080,14 @@ func (sc *StorageSmartContract) updateAllocationRequestInternal(
...
-	if request.Expiration < 0 &&
+	if newExpiration < 0 ||
 		newExpiration-t.CreationDate < toSeconds(conf.MinAllocDuration) {
 
 		return "", common.NewError("allocation_updating_failed",
 			"allocation duration becomes too short")
 	}
 
-	if request.Size < 0 && alloc.Size+request.Size < conf.MinAllocSize {
+	if alloc.Size+request.Size < conf.MinAllocSize || request.Size+alloc.Size < alloc.UsedSize {
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/allocation.go]: The `alloc.Size+request.Size` could be assigned to a variable and to make it cleaner.

**Review by peterlimg — CHANGES_REQUESTED** [review]

**Review by peterlimg — APPROVED** [review]
lgtm

**Pull request merged**